src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs

```diff
@@ -1285,21 +1285,149 @@ internal bool HasCollectionExpressionApplicableAddMethod(SyntaxNode syntax, Type
+                syntax,
+                right: syntax,
+                boundLeft: implicitReceiver,
+                rightName: methodName,
+                rightArity: 0,
+                typeArgumentsSyntax: default,
+                typeArgumentsWithAnnotations: default,
+                invoked: true,
+                indexed: false,
+                BindingDiagnosticBag.Discarded);
+            if (memberAccess is not BoundMethodGroup methodGroup)
+            {
+                return false;
+            }
+
+            Debug.Assert(methodGroup.SearchExtensionMethods);
+
+            if (methodGroup.ResultKind == LookupResultKind.Viable &&
+                methodGroup.Methods.Any(m => isApplicableAddMethod(m.GetConstructedLeastOverriddenMethod(accessingTypeOpt: null, requireSameReturnType: false), expectingExtensionMethod: false)))
+            {
+                return true;
+            }
+
+            if (methodGroup.SearchExtensionMethods &&
+                hasApplicableAddExtensionMethod(syntax, implicitReceiver))
+            {
+                return true;
+            }
+
+            return false;
+
+            bool hasApplicableAddExtensionMethod(SyntaxNode syntax, BoundObjectOrCollectionValuePlaceholder implicitReceiver)
+            {
+                foreach (var scope in new ExtensionMethodScopes(this))
+                {
+                    var methodGroup = MethodGroup.GetInstance();
+                    PopulateExtensionMethodsFromSingleBinder(scope, methodGroup, syntax, implicitReceiver, rightName: methodName, typeArgumentsWithAnnotations: default, BindingDiagnosticBag.Discarded);
+                    bool anyApplicable = methodGroup.Methods.Any(m => isApplicableAddMethod(m, expectingExtensionMethod: true));
+                    methodGroup.Free();
+                    if (anyApplicable)
+                    {
+                        return true;
+                    }
+                }
+                return false;
+            }
+
+            bool isApplicableAddMethod(MethodSymbol method, bool expectingExtensionMethod)
+            {
+                if (method.IsStatic != expectingExtensionMethod)
+                {
+                    return false;
+                }
+
+                var parameters = method.Parameters;
+                int valueIndex = expectingExtensionMethod ? 1 : 0;
+                int requiredLength = valueIndex + 1;
+                if (parameters.Length < requiredLength)
+                {
+                    return false;
+                }
+
+                // Any trailing parameters must be optional or params.
+                int optionalLength = parameters.Length - (OverloadResolution.IsValidParams(this, method) ? 1 : 0);
+                for (int i = requiredLength; i < optionalLength; i++)
+                {
+                    if (parameters[i] is not ({ IsOptional: true, RefKind: RefKind.None or RefKind.In or RefKind.RefReadOnlyParameter }))
```

Comment:
>RefKind: RefKind.None or RefKind.In or RefKind.RefReadOnlyParameter [](http://example.com/codeflow?start=66&length=67)

Could you add a comment pointing to the code that enforces this restriction during invocation binding?

Reply:
I think ```ParameterSymbol.IsOptional``` does that, and it looks like it allows `ref` optional in some scenarios